### PR TITLE
Allow octant to start without a kubeconfig

### LINKED ISSUE #362: Allow Octant to start without a kubeconfig.
Allow octant to start without a kubeconfig.

The UI should prompt the user for a path to a valid kubeconfig file.

Currently Octant immediately fails if unable to parse the kubeconfig file. We want to redirect that behavior to allow for a UI that would enable a user to select a kubeconfig.

A plugin should also be able to set the kubeconfig.

### LINKED ISSUE #750: Proper onboarding for new users
**What steps did you take and what happened:**

> brew install octant
> octant

🤮

> 2020-03-04T09:10:46.824+0100	ERROR	commands/dash.go:109	dashboard failed	{"err": "failed to init cluster client: invalid configuration: no configuration has been provided"}
github.com/vmware-tanzu/octant/internal/commands.newOctantCmd.func1.2
	/private/tmp/octant-20200215-54834-lxnkt8/src/github.com/vmware-tanzu/octant/internal/commands/dash.go:109

**What users expect**

To see the octant user interface or a proper help message indicating what they should do

**What most users do**

> brew remove octant

### PR DESCRIPTION
**Pull request by GuessWhoSamFoo**
**What this PR does / why we need it**:
- [x] Fix modal flicker
- [x] ~Add file browser~ To do in a future PR
- [x] ~Handle case where CRD watcher fails to start (valid kubeconfig, but connection refused)~ To do in a future PR

closes #750 

**Which issue(s) this PR fixes**
- Fixes #362

Signed-off-by: GuessWhoSamFoo <foos@vmware.com>

### CODE REVIEW

File: pkg/dash/dash.go [hunk 1]
```diff
@@ -458,3 +527,24 @@ func enableOpenCensus() error {
 
 	return nil
 }
+
+// findKubeConfig looks for kube config from .kube or provided by user
```
Comment (wwitzel3) [inline pkg/dash/dash.go:531]: Not sure this function is working as intended.

File: pkg/dash/dash.go [hunk 2]
```diff
@@ -458,3 +527,24 @@ func enableOpenCensus() error {
...
+func findKubeConfig(temporaryKubeConfig kubeconfig.TemporaryKubeConfig) <-chan string {
+	r := make(chan string)
+	go func() {
+		kubeconfig := clientcmd.NewDefaultClientConfigLoadingRules().GetDefaultFilename()
+		if _, err := os.Stat(kubeconfig); err == nil {
```
Comment (wwitzel3) [inline pkg/dash/dash.go:536]: os.Stat does not return an error if the file does not exist. So this function is likely just returning. I'd recommend writing some tests around this to assert the behavior.

File: pkg/dash/dash.go [hunk 3]
```diff
@@ -458,3 +527,24 @@ func enableOpenCensus() error {
...
+			r <- kubeconfig
+			return
+		}
+
+		defer close(r)
```
Comment (wwitzel3) [inline pkg/dash/dash.go:541]: I think you'll want to move this up to the start of the go routine.

**Review by wwitzel3 — CHANGES_REQUESTED** [review]
If I attempt to submit the empty uploader component, I get a validation error which is masked by the modal overlay (not a big deal). The dialog will let me attempt to submit an empty form.

I attempt to submit an invalid kubeconfig, then I paste in a valid kubeconfig and submit, but when the page reloads, I am still seeing the upload dialog and have to manually refresh to get the correct screen to load.

- Disable the upload button unless there is a value in the field.
- Fix the issue of a malformed kubeconfig, followed by a well formed one, not properly reloading Octant.
- Write some tests for findKubeConfig and assert it is doing the correct thing.

```
2020-06-29T15:50:00.840-0400    INFO    config/dash.go:230      updated kube config context     {"new-kube-context": ""}
2020-06-29T15:50:01.290-0400    INFO    dash/dash.go:156        using api service
2020-06-29T15:50:01.348-0400    INFO    config/dash.go:230      updated kube config context     {"new-kube-context": ""}```

File: pkg/dash/dash.go (hunk 2 above)
Comment (wwitzel3) [reply pkg/dash/dash.go:536]: I think what you want here is to use the fsnotify library and/or modify the existing kubeconfig watcher to expose an event you can use to write to your channel.

**Review by mklanjsek — COMMENTED** [review]
Spent some time playing with your changes. Overall it looks awesome, just a couple of comments:

- After pasting YAML and then later on restarting the Octant, I was confronted with the same initial screen prompting entering new cubeconfig. I expected that the cubeconfig info entered the first time around would be saved and reopened after restarting Octant.
- Is `UploaderComponent` something that will be exposed to plugin developers? In that case it would need a documentation story.

**Comment by GuessWhoSamFoo** [comment]
- Disabled submit button if input text is empty
 - Fixed some issues with modal appearing on reload cases

@mklanjsek We will never expose the `UploaderComponent` since it has handlers that rely on the backend to be in a certain state. A future component for browsing the filesystem can be added for plugin authors.

File: pkg/dash/dash.go (hunk 2 above)
Comment (wwitzel3) [reply pkg/dash/dash.go:538]: @GuessWhoSamFoo you are right, my comment about os.Stat not returning the error wasn't fully completed. What I was really trying to say here is that os.Stat isn't being called again. So even if it errors, when you do add a kubeconfig to that path the function is still just waiting for an upload. Not periodically checking for a new kubeconfig.

File: pkg/dash/dash.go [hunk 4]
```diff
@@ -60,21 +62,125 @@ type Options struct {
...
+					r.loadingComplete <- true
+					return
+				}
+			}
+		}
+	}()
+
+	go func() {
+		<-r.loadingComplete
+		apiService, err := r.initAPI(ctx, logger, options)
```
Comment (wwitzel3) [inline pkg/dash/dash.go]: Consider merging these into a single go routine. You are blocking on findKubeConfig already. This is the only writer to `r.loadingComplete` which means you could just as well have the contents of the second go routine as a function that gets called.

File: pkg/dash/dash.go (hunk 2 above)
Comment (GuessWhoSamFoo) [reply pkg/dash/dash.go:516]: This is now moved to `loading_state.go` because the watcher detecting a new config will also require a refresh event.

**Comment by mklanjsek** [comment]
I played with it again and now seeing some weirdness. After renaming my current config file and starting Octant, the `Add a new kubeconfig` popup shows up as expected but I am not able to go past that.  I tried entering different versions of kubeconfig files (including exact copy of my latest one) but after hitting the Upload button the popup just refreshes with empty content.

One tiny usability thing: `This can't be empty` field stays visible even after entering text kubeconfig field.

**Comment by GuessWhoSamFoo** [comment]
@mklanjsek Have you cleared cache? I also had to rebuild node_modules at some point due to the dep bumps and rebasing.

**Comment by mklanjsek** [comment]
> @mklanjsek Have you cleared cache? I also had to rebuild node_modules at some point due to the dep bumps and rebasing.

Must be that 🤦🏽‍♂️

I'll give it another try on master and let you know if I still see that.

**Comment by wwitzel3** [comment]
When I install a kubeconfig, this happens.

```
 ⚙  ~/repos/octant   GuessWhoSamFoo-kubeconfig-start  mv ~/.kube/bak-config ~/.kube/config
 ⚙  ~/repos/octant   GuessWhoSamFoo-kubeconfig-start  2020-07-09T12:25:23.962-0400  INFO    module/manager.go:79    registering action      {"component": "module-manager", "actionPath": "overview/startPortForward", "module-name": "overview"}
2020-07-09T12:25:23.962-0400    INFO    module/manager.go:79    registering action      {"component": "module-manager", "actionPath": "action.octant.dev/cordon", "module-name": "overview"}
2020-07-09T12:25:23.962-0400    INFO    module/manager.go:79    registering action      {"component": "module-manager", "actionPath": "action.octant.dev/update", "module-name": "overview"}
2020-07-09T12:25:23.962-0400    INFO    module/manager.go:79    registering action      {"component": "module-manager", "actionPath": "action.octant.dev/serviceEditor", "module-name": "overview"}
2020-07-09T12:25:23.962-0400    INFO    module/manager.go:79    registering action      {"component": "module-manager", "actionPath": "action.octant.dev/containerEditor", "module-name": "overview"}
2020-07-09T12:25:23.962-0400    INFO    module/manager.go:79    registering action      {"component": "module-manager", "actionPath": "overview/stopPortForward", "module-name": "overview"}
2020-07-09T12:25:23.962-0400    INFO    module/manager.go:79    registering action      {"component": "module-manager", "actionPath": "action.octant.dev/uncordon", "module-name": "overview"}
2020-07-09T12:25:23.963-0400    INFO    module/manager.go:79    registering action      {"component": "module-manager", "actionPath": "action.octant.dev/cronJob", "module-name": "overview"}
2020-07-09T12:25:23.963-0400    INFO    module/manager.go:79    registering action      {"component": "module-manager", "actionPath": "action.octant.dev/deploymentConfiguration", "module-name": "overview"}
2020-07-09T12:25:23.963-0400    INFO    module/manager.go:79    registering action      {"component": "module-manager", "actionPath": "action.octant.dev/deleteObject", "module-name": "configuration"}
2020-07-09T12:25:25.477-0400    INFO    config/dash.go:230      updated kube config context     {"new-kube-context": ""}
2020-07-09T12:25:25.970-0400    INFO    dash/dash.go:548        using api service
2020-07-09T12:25:26.031-0400    INFO    config/dash.go:230      updated kube config context     {"new-kube-context": ""}
panic: send on closed channel
        panic: close of closed channel

goroutine 454 [running]:
github.com/vmware-tanzu/octant/internal/api.(*LoadingManager).WatchConfig(0xc0007f2300, 0xc0002582a0, 0x3423020, 0xc0006b7080, 0x348aa40, 0x4213aa8)
        /home/wwitzel3/repos/octant/internal/api/loading_state.go:138 +0x112
created by github.com/vmware-tanzu/octant/internal/api.(*LoadingManager).Start
        /home/wwitzel3/repos/octant/internal/api/loading_state.go:66 +0xab

[2]  + 31413 exit 2     ./build/octant
```

File: pkg/dash/dash.go [hunk 5]
```diff
@@ -458,3 +503,48 @@ func enableOpenCensus() error {
...
+			r <- kubeconfig
+			return
+		}
+
+		select {
+		case path, ok := <-temporaryKubeConfig.Path:
+			if ok {
+				r <- path
+			}
+		}
```
Comment (wwitzel3) [inline pkg/dash/dash.go:527]: No need for the select here.

    path, ok := <-temporaryKubeConfig.Path
    if ok {
        r <- path
    }

File: pkg/dash/dash.go [hunk 6]
```diff
@@ -74,7 +78,85 @@ func NewRunner(ctx context.Context, logger log.Logger, options Options) (*Runner
...
+		if r.dash != nil {
+			select {
+			case options.KubeConfig = <-findKubeConfig(logger, r.dash.kubeconfig):
+				if options.KubeConfig != "" {
+					logger.Debugf("Loading configuration: %v", options.KubeConfig)
+					go r.startAPIService(ctx, logger, options)
+					return
+				}
+			}
+		}
```
Comment (wwitzel3) [inline pkg/dash/dash.go:133]: Another instance where you don't need the select.

File: internal/api/loading_state.go [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+
+	if err := tempFile.Close(); err != nil {
+		return err
+	}
+
+	l.client.Send(octant.Event{
+		Type: octant.EventTypeRefresh,
+	})
+
+	l.kubeConfig.Path <- tempFile.Name()
```
Comment (wwitzel3) [inline internal/api/loading_state.go:123]: I don't understand why the KubeConfig contents are sent in to a channel on TemporaryKubeConfig and then after we save the temporary file, we send the location in to the Path channel.

Do we need both?

File: pkg/dash/dash.go [hunk 8]
```diff
@@ -458,3 +503,48 @@ func enableOpenCensus() error {
 
 	return nil
 }
+
+// findKubeConfig looks for kube config from .kube or provided by user
+func findKubeConfig(logger log.Logger, temporaryKubeConfig kubeconfig.TemporaryKubeConfig) <-chan string {
+	r := make(chan string)
+
+	go func() {
```
Comment (wwitzel3) [inline pkg/dash/dash.go:511]: The one location where you call `findKubeConfig` is already a goroutine, you don't need the nested go routine here.

**Review by wwitzel3 — CHANGES_REQUESTED** [review]
This is looking great, final set of comments and then this is ready to merge.

**Comment by wwitzel3** [comment]
If I have a valid kubeconfig installed, I should not see the loading screen when launching Octant.

loading-screen-always

**Review by wwitzel3 — APPROVED** [review]

**Pull request merged**